src/PowerShellEditorServices/Language/TokenOperations.cs

```diff
@@ -21,103 +21,120 @@ internal static class TokenOperations
-            foldableRegions.AddRange(
-                MatchTokenElements(tokens, s_openingBraces, TokenKind.RCurly, RegionKindNone)
-            );
+            var refList = new FoldingReferenceList();
 
-            // Find matching parentheses     ( -> )
-            // Find matching array literals @( -> )
-            // Find matching subexpressions $( -> )
-            foldableRegions.AddRange(
-                MatchTokenElements(tokens, s_openingParens, TokenKind.RParen, RegionKindNone)
-            );
-
-            // Find contiguous here strings @' -> '@
-            foldableRegions.AddRange(
-                MatchTokenElement(tokens, TokenKind.HereStringLiteral, RegionKindNone)
-            );
-
-            // Find unopinionated variable names ${ \n \n }
-            foldableRegions.AddRange(
-                MatchTokenElement(tokens, TokenKind.Variable, RegionKindNone)
-            );
-
-            // Find contiguous here strings @" -> "@
-            foldableRegions.AddRange(
-                MatchTokenElement(tokens, TokenKind.HereStringExpandable, RegionKindNone)
-            );
+            Stack<Token> tokenCurlyStack = new Stack<Token>();
+            Stack<Token> tokenParenStack = new Stack<Token>();
+            foreach (Token token in tokens)
+            {
+                // Find matching braces  { -> }
+                // Find matching hashes @{ -> }
+                if ((token.Kind == TokenKind.LCurly) || (token.Kind == TokenKind.AtCurly)) {
+                    tokenCurlyStack.Push(token);
+                }
+                if ((tokenCurlyStack.Count > 0) && (token.Kind == TokenKind.RCurly)) {
+                    refList.SafeAdd(CreateFoldingReference(tokenCurlyStack.Pop(), token, RegionKindNone));
+                }
+                // Find matching parentheses     ( -> )
+                // Find matching array literals @( -> )
+                // Find matching subexpressions $( -> )
+                if ((token.Kind == TokenKind.LParen) || (token.Kind == TokenKind.AtParen) || (token.Kind == TokenKind.DollarParen)) {
+                    tokenParenStack.Push(token);
+                }
+                if ((tokenParenStack.Count > 0) && (token.Kind == TokenKind.RParen)) {
+                    refList.SafeAdd(CreateFoldingReference(tokenParenStack.Pop(), token, RegionKindNone));
+                }
+                // Find contiguous here strings @' -> '@
+                // Find unopinionated variable names ${ \n \n }
+                // Find contiguous expandable here strings @" -> "@
+                if ((
+                    (token.Kind == TokenKind.HereStringLiteral) ||
+                    (token.Kind == TokenKind.Variable) ||
+                    (token.Kind == TokenKind.HereStringExpandable)
+                    ) && (token.Extent.StartLineNumber != token.Extent.EndLineNumber)) {
+                    refList.SafeAdd(CreateFoldingReference(token, token, RegionKindNone));
+                }
+            }
```

Comment:
I think a switch might be slightly more readable here:

```csharp
            switch (token.Kind)
            {
                case LCurly:
                case AtCurly:
                    tokenCurlyStack.Push(token);
                    break;

                case RCurly:
                    if (tokenCurlyStack.Count > 0)
                    {
                        refList.SafeAdd(CreateFoldingReference(tokenCurlyStack.Pop(), token, RegionKindNone));
                    }
                    break;

                case LParen:
                case AtParen:
                case DollarParen:
                    tokenParenStack.Push(token);
                    break;

                case RParen:
                    if (tokenParenStack.Count > 0)
                    {
                        refList.SafeAdd(CreateFoldingReference(tokenParenStack.Pop(), token, RegionKindNone));
                    }
                    break;

                case HereStringLiteral:
                case Variable:
                case HereStringExpandable:
                    if (token.Extent.StartLineNumber != token.Extent.EndLineNumber)
                    {
                        refList.SafeAdd(CreateFoldingReference(token, token, RegionKindNone));
                    }
                    break;

                default:
                    break;
            }
```

Reply:
Forgot about those.  Changed.